opentelemetry-api/src/opentelemetry/trace/__init__.py

```diff
@@ -102,34 +143,44 @@ def get_context(self) -> "SpanContext":
             A :class:`.SpanContext` with a copy of this span's immutable state.
         """
 
-    def set_attribute(
-        self: "Span", key: str, value: types.AttributeValue
-    ) -> None:
+    def set_attribute(self, key: str, value: types.AttributeValue) -> None:
         """Sets an Attribute.
 
         Sets a single Attribute with the key and value passed as arguments.
         """
 
     def add_event(
-        self: "Span", name: str, attributes: types.Attributes = None
+        self, name: str, attributes: types.Attributes = None
     ) -> None:
-        """Adds an Event.
+        """Adds an `Event`.
 
-        Adds a single Event with the name and, optionally, attributes passed
+        Adds a single `Event` with the name and, optionally, attributes passed
         as arguments.
         """
 
+    def add_lazy_event(self, event: Event) -> None:
+        """Adds an `Event`.
+
+        Adds an `Event` that has previously been created.
+        """
+
     def add_link(
-        self: "Span",
+        self,
         link_target_context: "SpanContext",
         attributes: types.Attributes = None,
     ) -> None:
-        """Adds a Link to another span.
+        """Adds a `Link` to another span.
 
-        Adds a single Link from this Span to another Span identified by the
+        Adds a single `Link` from this Span to another Span identified by the
         `SpanContext` passed as argument.
         """
 
+    def add_lazy_link(self, link: "Link") -> None:
```

Comment:
I would almost argue that add_link should just always take a link object. Is it that much harder to do:

```
span.add_link(context, [attribute_1])
````
instead of
```
span.add_link(Link(context, [attribute_1]))
```

Reply:
Yeah, it seems a bit redundant as of now, though I'd vote for having only `span.add_link(context, [attribute_1])`. I don't think I fully grasp the design of these "lazy" method yet, I can't see what's lazy about them.

Reply:
I don't have a strong opinion on any of both options, I just followed the specs: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/api-tracing.md#add-links, there the two methods are required `add_xxx` that receives the parameters and  `add_lazy_xxx` that receives an object.

Reply:
We could optionally have only the `add_link(link: Link)` method, to reduce API surface - but we won't *easily* be able in the future to add an overload that takes a `SpanContext` + attributes, because of the name (`add_nonlazy_link()`?)

For historical reference: we added these as in Java this is a usual thing to add there (overload with different parameters, that is). We could drop this for Python if we feel like that, I feel.